src/Analyser/TypeSpecifier.php

```diff
@@ -501,6 +503,8 @@ public function specifyTypesInCondition(
 					$argType = $scope->getType($expr->right->getArgs()[0]->value);
 					if ($argType instanceof StringType) {
 						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new AccessoryNonEmptyStringType(), $context, false, $scope));
+					} else {
+						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new ConstantStringType(''), $context->negate(), false, $scope));
```

Comment:
This fixes `is_string($a) && strlen($a) > 0 || null === $a`. The problem seems to be that in that case `$argType` is still `MixedType` here. I'm not sure why, but this smells a bit like a hacky workaround? If so, I'm fine with removing it and the relevant assertions in order to fix the rest for now. I was not successful debugging everything that is going on leading to `MutatingScope::$resolvedTypes` being set incorrectly..

Reply:
I think there are a couple of interesting parts here and in #1012 
- the type specification for the left part for `is_string` is coming from a TypeSpecifyingExtension
- the right part does not know about the type from the left part and if it asks the scope for it, it gets mixed. this is sometimes a problem, sometimes not. it looks though like here it can still find a type for `strlen($a) > 0` in the right part
![image](https://user-images.githubusercontent.com/5738896/153940688-7badb8a6-df1e-4bbb-9d5f-90500b779be2.png)
the scope is using `filterBySpecifiedTypes` afterwards, and then has the right type (non-empty-string) if asked for the expression or the variable type
- but in this case, there is also the BooleanOr and this is where, again, type information (the `strlen($a) > 0`) is lost via intersection of 2 SpecifiedTypes instances

Reply:
OK, I came up with something, resolving this for now and adding a new comment down below